Add python bindings for an HNSW index fixture

toregge opened the pull request:
@geirst : please review

toregge committed:
Add python bindings for a HNSW index fixture using a tensor attribute vector containing a nearest neighbor index.

geirst reviewed (approved):
This looks very good 👍 Some comments and suggestions to consider.

geirst commented:
I think @aressem and / or @baldersheim should have a look at the module, rpm, and python binding changes.

geirst commented:
@jobergum FYI

toregge committed:
Updates based on review feedback:

Remove distance_threshold argument to find_top_k member function in
HNSW index fixture class.

Remove duplicate tests in unit test.

Adjust comments and descriptions.

toregge commented on ann_benchmark/src/tests/ann_benchmark/test_euclidean.py:
```diff
@@ -0,0 +1,63 @@
+    f = Fixture()
+    f.set(0, [0, 0])
+    f.set(1, [10, 10])
+    assert f.get(0) == [0, 0]
```
Removed.

toregge commented on ann_benchmark/src/vespa/ann_benchmark/vespa_ann_benchmark.cpp:
```diff
@@ -0,0 +1,217 @@
+}
+
+/*
+ * Wrapper class for a tensor attribute vector containing a nearest
```
Suggested change applied.

toregge commented on ann_benchmark/src/vespa/ann_benchmark/vespa_ann_benchmark.cpp:
```diff
@@ -0,0 +1,217 @@
+
+/*
+ * Wrapper class for a tensor attribute vector containing a nearest
+ * neighbor index.
```
Suggested change applied.

toregge commented on ann_benchmark/src/vespa/ann_benchmark/vespa_ann_benchmark.cpp:
```diff
@@ -0,0 +1,217 @@
+    assert(_tensor_type.is_dense());
+    assert(_tensor_type.count_indexed_dimensions() == 1u);
+    _dim_size = _tensor_type.dimensions()[0].size;
+    std::cout << "HnswIndex::HnswIndex Dimension size is " << _dim_size << std::endl;
```
Removed

toregge commented on ann_benchmark/src/vespa/ann_benchmark/vespa_ann_benchmark.cpp:
```diff
@@ -0,0 +1,217 @@
+}
+
+TopKResult
+HnswIndex::find_top_k(uint32_t k, const std::vector<float>& value, uint32_t explore_k, double distance_threshold)
```
Argument has been removed.

toregge commented on dist/vespa.spec:
```diff
@@ -437,6 +437,27 @@ Requires: %{name}-base-libs = %{version}-%{release}
+
+Vespa - The open big data serving engine - ann-benchmark
+
+Python binding for an HNSW index fixture using tensor attribute.
```
Suggested change applied.

toregge commented on ann_benchmark/src/vespa/ann_benchmark/setup.py:
```diff
@@ -0,0 +1,27 @@
+  version="0.1.0",
+  author="Tor Egge",
+  author_email="Tor.Egge@verizonmedia.com",
+  description="Python binding for an HNSW index fixture using tensor attribute",
```
Suggested change applied

geirst reviewed (approved):
👍

aressem commented on dist/vespa.spec:
```diff
@@ -437,6 +437,28 @@ Requires: %{name}-base-libs = %{version}-%{release}
 
 Vespa - The open big data serving engine - tools
 
+%package ann-benchmark
```
This seems very specific. Do we expect to add other benchmarks as well ? In this case we should consider just having vespa-benchmarks as a package.

toregge committed:
Don't use valgrind when unit testing python bindings.
